### ISSUE TITLE
`datalad get .` fails on first try

### ISSUE THREAD

**Issue by loj** [datalad/datalad#4209]
#### What is the problem?

When doing a `datalad get .` in an OpenNeuro dataset, it always fails to download some files the first time, but succeeds on subsequent `datalad get .` calls.

#### What steps will reproduce the problem?

<details><summary>First try fails to get all files.</summary>
<p>

```
❱ datalad clone https://github.com/OpenNeuroDatasets/ds000115.git tmpds
[INFO   ]   Remote origin not usable by git-annex; setting annex-ignore                                                                                                                        
[INFO   ] access to 1 dataset sibling s3-PRIVATE not auto-enabled, enable with:                                                                                                                
| 		datalad siblings -d "/home/loj/tmpds" enable -s s3-PRIVATE 
install(ok): /home/loj/tmpds (dataset)

❱ cd tmpds 

❱ datalad get .
Total (183 ok, 213 failed out of 396):  
[WARNING] Running get resulted in stderr output: git-annex: get: 213 failed 
[ERROR  ] Remote is configured to use versioning, but no S3 version ID is recorded for this key
| unknown export location [get(/home/loj/tmpds/sub-02/func/sub-02_task-letter2backtask_bold.nii.gz)] 
get(error): sub-02/func/sub-02_task-letter2backtask_bold.nii.gz (file) [Remote is configured to use versioning, but no S3 version ID is recorded for this key
unknown export location]

[...]

get(ok): sub-97/func/sub-97_task-letter2backtask_bold.nii.gz (file) [Remote is configured to use versioning, but no S3 version ID is recorded for this key]
action summary:
  get (error: 213, ok: 183)
```
</p>
</details>

<details><summary>Second try succeeds</summary>
<p>

```
❱ datalad get .
[...]
get(ok): sub-53/func/sub-53_task-letter0backtask_bold.nii.gz (file) [Remote is configured to use versioning, but no S3 version ID is recorded for this key]
action summary:
  get (ok: 213)
```
</p>
</details>

Limiting with `--jobs 2` has no impact.

Furthermore, if I ctrl. C immediately after launching the first `datalad get .` (before it has downloaded any files) then all files download successfully with a second `datalad get .`. 

#### What version of DataLad are you using? 

<details><summary>datalad wtf</summary>
<p>

```
❱ datalad wtf                                                                                                                                                                            
# WTF
## configuration <SENSITIVE, report disabled by configuration>
## datalad 
  - full_version: 0.12.2
  - version: 0.12.2
## dataset 
  - id: 381685ac-870b-11e8-8e19-0242ac12001c
  - metadata: <SENSITIVE, report disabled by configuration>
  - path: /home/loj/tmpds
  - repo: AnnexRepo
## dependencies 
  - appdirs: 1.4.3
  - boto: 2.44.0
  - cmd:7z: 16.02
  - cmd:annex: 7.20190819+git2-g908476a9b-1~ndall+1
  - cmd:bundled-git: 2.20.1
  - cmd:git: 2.20.1
  - cmd:system-git: 2.24.0
  - cmd:system-ssh: 7.9p1
  - exifread: 2.1.2
  - git: 3.0.5
  - gitdb: 2.0.6
  - humanize: 0.5.1
  - iso8601: 0.1.11
  - keyring: 17.1.1
  - keyrings.alt: 3.1.1
  - msgpack: 0.5.6
  - mutagen: 1.40.0
  - requests: 2.21.0
  - tqdm: 4.28.1
  - wrapt: 1.10.11
## environment 
  - GIT_PYTHON_GIT_EXECUTABLE: /usr/lib/git-annex.linux/git
  - LANG: en_US.UTF-8
  - LANGUAGE: en_US.UTF-8
  - LC_ALL: en_US.UTF-8
  - LC_CTYPE: en_US.UTF-8
  - PATH: /home/loj/.dotfiles/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/games
## extensions 
  - inm7: 
    - description: INM7 utilities
    - entrypoints: 
    - load_error: None
    - module: inm7_datalad
    - version: 0.5+10.g5f93ec5
  - ria: 
    - description: Helper for the remote indexed archive (RIA) special remote
    - entrypoints: 
      - ria_remote.export_archive.ExportArchive: 
        - class: ExportArchive
        - load_error: None
        - module: ria_remote.export_archive
        - names: 
          - ria-export-archive
          - ria_export_archive
    - load_error: None
    - module: ria_remote
    - version: 0.6+6.ge755961
## git-annex 
  - build flags: 
    - Assistant
    - Webapp
    - Pairing
    - S3
    - WebDAV
    - Inotify
    - DBus
    - DesktopNotify
    - TorrentParser
    - MagicMime
    - Feeds
    - Testsuite
  - dependency versions: 
    - aws-0.20
    - bloomfilter-2.0.1.0
    - cryptonite-0.25
    - DAV-1.3.3
    - feed-1.0.0.0
    - ghc-8.4.4
    - http-client-0.5.13.1
    - persistent-sqlite-2.8.2
    - torrent-10000.1.1
    - uuid-1.3.13
    - yesod-1.6.0
  - key/value backends: 
    - SHA256E
    - SHA256
    - SHA512E
    - SHA512
    - SHA224E
    - SHA224
    - SHA384E
    - SHA384
    - SHA3_256E
    - SHA3_256
    - SHA3_512E
    - SHA3_512
    - SHA3_224E
    - SHA3_224
    - SHA3_384E
    - SHA3_384
    - SKEIN256E
    - SKEIN256
    - SKEIN512E
    - SKEIN512
    - BLAKE2B256E
    - BLAKE2B256
    - BLAKE2B512E
    - BLAKE2B512
    - BLAKE2B160E
    - BLAKE2B160
    - BLAKE2B224E
    - BLAKE2B224
    - BLAKE2B384E
    - BLAKE2B384
    - BLAKE2BP512E
    - BLAKE2BP512
    - BLAKE2S256E
    - BLAKE2S256
    - BLAKE2S160E
    - BLAKE2S160
    - BLAKE2S224E
    - BLAKE2S224
    - BLAKE2SP256E
    - BLAKE2SP256
    - BLAKE2SP224E
    - BLAKE2SP224
    - SHA1E
    - SHA1
    - MD5E
    - MD5
    - WORM
    - URL
  - local repository version: 5
  - operating system: linux x86_64
  - remote types: 
    - git
    - gcrypt
    - p2p
    - S3
    - bup
    - directory
    - rsync
    - web
    - bittorrent
    - webdav
    - adb
    - tahoe
    - glacier
    - ddar
    - git-lfs
    - hook
    - external
  - supported repository versions: 
    - 5
    - 7
  - upgrade supported from repository versions: 
    - 0
    - 1
    - 2
    - 3
    - 4
    - 5
    - 6
  - version: 7.20190819+git2-g908476a9b-1~ndall+1
## location 
  - path: /home/loj/tmpds
  - type: dataset
## metadata_extractors 
  - annex: 
    - load_error: None
    - module: datalad.metadata.extractors.annex
    - version: None
  - audio: 
    - load_error: None
    - module: datalad.metadata.extractors.audio
    - version: None
  - datacite: 
    - load_error: None
    - module: datalad.metadata.extractors.datacite
    - version: None
  - datalad_core: 
    - load_error: None
    - module: datalad.metadata.extractors.datalad_core
    - version: None
  - datalad_rfc822: 
    - load_error: None
    - module: datalad.metadata.extractors.datalad_rfc822
    - version: None
  - exif: 
    - load_error: None
    - module: datalad.metadata.extractors.exif
    - version: None
  - frictionless_datapackage: 
    - load_error: None
    - module: datalad.metadata.extractors.frictionless_datapackage
    - version: None
  - image: 
    - load_error: None
    - module: datalad.metadata.extractors.image
    - version: None
  - xmp: 
    - load_error: No module named 'libxmp' [xmp.py:<module>:20]
    - module: datalad.metadata.extractors.xmp
## python 
  - implementation: CPython
  - version: 3.7.3
## system 
  - distribution: debian/10.3
  - encoding: 
    - default: utf-8
    - filesystem: utf-8
    - locale.prefered: UTF-8
  - max_path_length: 271
  - name: Linux
  - release: 4.19.0-8-amd64
  - type: posix
  - version: #1 SMP Debian 4.19.98-1 (2020-01-26)

```

</p>
</details>

**Comment by loj** [comment]
I ran this a couple more times and found that the behavior I described above doesn't completely hold true.

Subsequent `datalad get` calls also sometimes fail to get 1 or 2 files. But it's never the same level of failure as the initial `datalad get` call.

**Comment by yarikoptic** [comment]
oh, was I too quick to complain to openneuro folks?: https://github.com/OpenNeuroOrg/datalad-service/issues/79#issuecomment-591524075
I based my decision on observing "Remote is configured to use versioning, but no S3 version ID is recorded for this key unknown export location".  Will try now myself as well and look inside `git-annex` branch if aforementioned message is adequate.

**Comment by yarikoptic** [comment]
I am afraid it is a [known and not yet addressed issue in git-annex](https://git-annex.branchable.com/bugs/fails_to___96__get__96___in_parallel_for_a_freshly_clone_from_public_s3_bucket_where_versioning_info_was_forgotten/).  I will follow up there

<details>
<summary>Here is my protocol of struggle:</summary> 

```shell
(git-annex)lena:/tmp/tmpds[master]
$> sudo rm -rf tmpds && git clone https://github.com/OpenNeuroDatasets/ds000115.git tmpds && cd tmpds && git annex enableremote s3-PUBLIC && git annex get --json -J8 sub-01
Cloning into 'tmpds'...
remote: Enumerating objects: 4222, done.
remote: Counting objects: 100% (4222/4222), done.
remote: Compressing objects: 100% (2921/2921), done.
remote: Total 4222 (delta 259), reused 4222 (delta 259), pack-reused 0
Receiving objects: 100% (4222/4222), 389.19 KiB | 537.00 KiB/s, done.
Resolving deltas: 100% (259/259), done.
(merging origin/git-annex into git-annex...)
(recording state in git...)
(scanning for unlocked files...)

  Remote origin not usable by git-annex; setting annex-ignore
enableremote s3-PUBLIC ok
(recording state in git...)
git annex enableremote s3-PUBLIC  9.01s user 3.56s system 101% cpu 12.436 total
  Remote is configured to use versioning, but no S3 version ID is recorded for this key
  Remote is configured to use versioning, but no S3 version ID is recorded for this key
  Remote is configured to use versioning, but no S3 version ID is recorded for this key
  Remote is configured to use versioning, but no S3 version ID is recorded for this key
  unknown export location
  unknown export location
  unknown export location
{"command":"get","wanted":[{"here":false,"uuid":"3970e64d-b24a-48e7-83b4-1ec6fdaa229d","description":"[s3-PUBLIC]"},{"here":false,"uuid":"4a13d847-362b-4211-9124-861b9a8f4c13","description":"root@6dedfdc60b41:/datalad/ds000115"}],"note":"from s3-PUBLIC...\nUnable to access these remotes: s3-PUBLIC\nTry making some of these repositories available:\n\t3970e64d-b24a-48e7-83b4-1ec6fdaa229d -- [s3-PUBLIC]\n \t4a13d847-362b-4211-9124-861b9a8f4c13 -- root@6dedfdc60b41:/datalad/ds000115\n\n(Note that these git remotes have annex-ignore set: origin)","skipped":[],"success":false,"key":"MD5E-s19915954--e9f91cc343144c541ca35831c62e9503.nii.gz","file":"sub-01/func/sub-01_task-letter1backtask_bold.nii.gz"}
{"command":"get","wanted":[{"here":false,"uuid":"3970e64d-b24a-48e7-83b4-1ec6fdaa229d","description":"[s3-PUBLIC]"},{"here":false,"uuid":"4a13d847-362b-4211-9124-861b9a8f4c13","description":"root@6dedfdc60b41:/datalad/ds000115"}],"note":"from s3-PUBLIC...\nUnable to access these remotes: s3-PUBLIC\nTry making some of these repositories available:\n\t3970e64d-b24a-48e7-83b4-1ec6fdaa229d -- [s3-PUBLIC]\n \t4a13d847-362b-4211-9124-861b9a8f4c13 -- root@6dedfdc60b41:/datalad/ds000115\n\n(Note that these git remotes have annex-ignore set: origin)","skipped":[],"success":false,"key":"MD5E-s19916325--b0bc24ff278e3ef7031e6c801fc6c694.nii.gz","file":"sub-01/func/sub-01_task-letter2backtask_bold.nii.gz"}
{"command":"get","wanted":[{"here":false,"uuid":"3970e64d-b24a-48e7-83b4-1ec6fdaa229d","description":"[s3-PUBLIC]"},{"here":false,"uuid":"4a13d847-362b-4211-9124-861b9a8f4c13","description":"root@6dedfdc60b41:/datalad/ds000115"}],"note":"from s3-PUBLIC...\nUnable to access these remotes: s3-PUBLIC\nTry making some of these repositories available:\n\t3970e64d-b24a-48e7-83b4-1ec6fdaa229d -- [s3-PUBLIC]\n \t4a13d847-362b-4211-9124-861b9a8f4c13 -- root@6dedfdc60b41:/datalad/ds000115\n\n(Note that these git remotes have annex-ignore set: origin)","skipped":[],"success":false,"key":"MD5E-s35278866--09ae0c4f8a8eb6d379104e9ab9ca577a.nii.gz","file":"sub-01/anat/sub-01_T1w.nii.gz"}
{"command":"get","note":"from s3-PUBLIC...\nchecksum...","success":true,"key":"MD5E-s19875519--1a412c009b3d247b95f774b9099d55a8.nii.gz","file":"sub-01/func/sub-01_task-letter0backtask_bold.nii.gz"}
git-annex: get: 3 failed
(dev3) 1 30077 ->1.....................................:Wed 26 Feb 2020 12:04:41 PM EST:.
(git)lena:/tmp/tmpds/tmpds[master]
$> git annex get --json -J8 sub-01
  Remote is configured to use versioning, but no S3 version ID is recorded for this key
  Remote is configured to use versioning, but no S3 version ID is recorded for this key
  Remote is configured to use versioning, but no S3 version ID is recorded for this key
{"command":"get","note":"from s3-PUBLIC...\nchecksum...","success":true,"key":"MD5E-s19915954--e9f91cc343144c541ca35831c62e9503.nii.gz","file":"sub-01/func/sub-01_task-letter1backtask_bold.nii.gz"}
{"command":"get","note":"from s3-PUBLIC...\nchecksum...","success":true,"key":"MD5E-s19916325--b0bc24ff278e3ef7031e6c801fc6c694.nii.gz","file":"sub-01/func/sub-01_task-letter2backtask_bold.nii.gz"}
{"command":"get","note":"from s3-PUBLIC...\nchecksum...","success":true,"key":"MD5E-s35278866--09ae0c4f8a8eb6d379104e9ab9ca577a.nii.gz","file":"sub-01/anat/sub-01_T1w.nii.gz"}
(dev3) 1 30078.....................................:Wed 26 Feb 2020 12:07:07 PM EST:.
(git)lena:/tmp/tmpds/tmpds[master]
$> ls -l sub-01/anat/sub-01_T1w.nii.gz
lrwxrwxrwx 1 yoh yoh 142 Feb 26 12:04 sub-01/anat/sub-01_T1w.nii.gz -> ../../.git/annex/objects/gq/Mq/MD5E-s35278866--09ae0c4f8a8eb6d379104e9ab9ca577a.nii.gz/MD5E-s35278866--09ae0c4f8a8eb6d379104e9ab9ca577a.nii.gz
(dev3) 1 30079.....................................:Wed 26 Feb 2020 12:07:20 PM EST:.
(git)lena:/tmp/tmpds/tmpds[master]
$> git annex version
git-annex version: 7.20200204+git62-gcc4521068-1~ndall+1
build flags: Assistant Webapp Pairing S3 WebDAV Inotify DBus DesktopNotify TorrentParser MagicMime Feeds Testsuite
dependency versions: aws-0.20 bloomfilter-2.0.1.0 cryptonite-0.25 DAV-1.3.3 feed-1.0.0.0 ghc-8.4.4 http-client-0.5.13.1 persistent-sqlite-2.8.2 torrent-10000.1.1 uuid-1.3.13 yesod-1.6.0
key/value backends: SHA256E SHA256 SHA512E SHA512 SHA224E SHA224 SHA384E SHA384 SHA3_256E SHA3_256 SHA3_512E SHA3_512 SHA3_224E SHA3_224 SHA3_384E SHA3_384 SKEIN256E SKEIN256 SKEIN512E SKEIN512 BLAKE2B256E BLAKE2B256 BLAKE2B512E BLAKE2B512 BLAKE2B160E BLAKE2B160 BLAKE2B224E BLAKE2B224 BLAKE2B384E BLAKE2B384 BLAKE2BP512E BLAKE2BP512 BLAKE2S256E BLAKE2S256 BLAKE2S160E BLAKE2S160 BLAKE2S224E BLAKE2S224 BLAKE2SP256E BLAKE2SP256 BLAKE2SP224E BLAKE2SP224 SHA1E SHA1 MD5E MD5 WORM URL
remote types: git gcrypt p2p S3 bup directory rsync web bittorrent webdav adb tahoe glacier ddar git-lfs hook external
operating system: linux x86_64
supported repository versions: 7
upgrade supported from repository versions: 0 1 2 3 4 5 6
local repository version: 7
```
</details>
Interestingly, on another server with ample bandwidth and older git annex 7.20190819+git2-g908476a9b-1~ndall+1 -- it managed to get all the files just fine (well -- with all the complaints about missing history info).

**Comment by yarikoptic** [comment]
ok, [troubleshooted and followed up](https://git-annex.branchable.com/bugs/fails_to___96__get__96___in_parallel_for_a_freshly_clone_from_public_s3_bucket_where_versioning_info_was_forgotten/#comment-77e82bc6e3f7f29aa895f5777a6a3879).  @loj please try running `datalad -c annex.retry=2 get .` . IMHO it is just a workaround. 
Meanwhile it reminded me about https://github.com/datalad/datalad/issues/2896 which I never acted upon.  Although implementing it would only mask out possible issues with git annex, I think we are doomed to do it.

**Comment by yarikoptic** [comment]
FWIW debian-devel of neurodebian now has 7.20200219+git135-g34d726cba-1~ndall+1 which carries the fix for this issue.  I will close this issue - workaround is stated, fixed annex is available, and may be #2896 would also be implemented.

**Comment by loj** [comment]
Thanks for looking into this @yarikoptic :-)

I tried running `datalad -c annex.retry=2 get .`, and got the same behavior as before.

<details><summary>With annex.retry.</summary>
<p>

```
❱ datalad clone https://github.com/OpenNeuroDatasets/ds000115.git tmpds
[INFO   ]   Remote origin not usable by git-annex; setting annex-ignore                                                                                                                        
[INFO   ] access to 1 dataset sibling s3-PRIVATE not auto-enabled, enable with:                                                                                                                
| 		datalad siblings -d "/home/loj/tmpds" enable -s s3-PRIVATE 
install(ok): /home/loj/tmpds (dataset)

❱ cd tmpds

❱ datalad -c annex.retry=2 get .         
Total (197 ok, 199 failed out of 396):
[WARNING] Running get resulted in stderr output: git-annex: get: 199 failed 
[ERROR  ] Remote is configured to use versioning, but no S3 version ID is recorded for this key
| unknown export location [get(/home/loj/tmpds/sub-01/anat/sub-01_T1w.nii.gz)] 
get(error): sub-01/anat/sub-01_T1w.nii.gz (file) [Remote is configured to use versioning, but no S3 version ID is recorded for this key
unknown export location]

[...]

get(ok): sub-99/func/sub-99_task-letter1backtask_bold.nii.gz (file) [Remote is configured to use versioning, but no S3 version ID is recorded for this key]
action summary:
  get (error: 199, ok: 197)

❱ datalad -c annex.retry=2 get .
[...]
get(ok): sub-50/anat/sub-50_T1w.nii.gz (file) [Remote is configured to use versioning, but no S3 version ID is recorded for this key]
action summary:
  get (ok: 199)
```
</p>
</details>

I also tried increasing the number of retries, but that had no effect.

**Comment by mih** [comment]
@loj was this with or without the annex development version that @yarikoptic mentioned?

**Comment by loj** [comment]
It was without.

**Comment by yarikoptic** [comment]
```
❱ datalad -c annex.retry=2 get .         
Total (197 ok, 199 failed out of 396):
[WARNING] Running get resulted in stderr output: git-annex: get: 199 failed 
```
hm... so `annex.retry` is not sufficient workaround - I hoped it would, since it was for me ... or may be needs higher (than 2) number in your case.  might be worth checking (but you would need to downgrade annex first)

**Comment by yarikoptic** [comment]
BUT that version ` 7.20200219+git135-g34d726cba-1~ndall+1` generally must not be used, since it is not compatible with any datalad in existence at the moment, and also what to be v8 annex -- with all the repos autoupgrades (unless disabled) etc.  I guess I could have tried to patch the one we have in neurodebian with the fix (if applicable) but I do not think that would be a scalable solution since none other distributed annex would be.

**Comment by yarikoptic** [comment]
I will reopen this issue since we still have an issue which either would get addressed by fixing #4185, or some other workaround.

**Comment by yarikoptic** [comment]
I will just assume that it was addressed by #4185 ... if not -- feel welcome to reopen again ;-)
